### ISSUE TITLE
week3

### ISSUE THREAD

**Issue by jioxia** [SXueLing/BadouNLP#1663]
import torch
import torch.nn as nn
import numpy as np
import random
import json
import matplotlib.pyplot as plt

"""

基于pytorch框架编写模型训练
实现一个自行构建的机器学习任务
x是一个五维向量，如果第一个数>第五个数，则为正样本，反之为负样本

"""

class torch_model(nn.Module):
    def __init__(self, input_size):
        super(torch_model, self).__init__()
        self.liner = nn.RNN(input_size, 5)
        self.activation = torch.sigmoid
        self.loss = nn.functional.cross_entropy  # 均方差loss/交叉熵loss

    # 当输入真实标签，返回loss值，无真实标签，返回预测值
    def forward(self, x, y = None):
        # 添加时间步维度
        x = x.unsqueeze(1)  # 形状变为 (batch_size, 1, input_size)
        x, _ = self.liner(x)  # 只取RNN的输出序列
        y_pred = self.activation(x[:, -1, :])  # 取最后一个时间步的输出
        if y is not None:
            return self.loss(y_pred, y.long())  # 标签需要是长整型
        else:
            return y_pred

# 随机生成一个5维向量，如果第一个值大于第五个，认为是正样本，反之为负样本
# 如果第一个数加第二个数大于第三个数为第一类，第二个数加第三个数大于第四个数为第二类
# 第三个数加第四个数大于第五个数
def build_sample():
    x = np.random.random(5)
    if x[0] + x[1] > x[2]:
        return x, 1
    elif x[1] + x[2] > x[3]:
        return x, 2
    elif x[2] + x[3] > x[4]:
        return x, 3
    else:
        return x, 0

# 随机生成一批正负样本均匀的样本
def build_dataset(total_sample_num):
    X = []
    Y = []
    for i in range(total_sample_num):
        x, y = build_sample()
        X.append(x)
        Y.append(y)
    # 先转换为NumPy数组，再转换为张量
    return torch.FloatTensor(X), torch.LongTensor(Y)

# 测试代码，用来测试每轮模型的准确率
def evaluate(model):
    model.eval()
    test_sample_num = 100
    x, y = build_dataset(test_sample_num)
    count1 = (y == 1).sum().item()
    count2 = (y == 2).sum().item()
    count3 = (y == 3).sum().item()
    count0 = (y == 0).sum().item()
    print("本次测试集第一类有%d个，第二类有%d个，第三类有%d个，第0类有%d个" % (count1, count2, count3, count0))

    correct, wrong = 0, 0
    with torch.no_grad():
        y_pred = model(x)  # 模型预测 model.forward(x)
        y_pred = torch.argmax(y_pred, dim=1)  # 取最大值的索引作为预测类别
        for y_p, y_t in zip(y_pred, y):
            if y_p == int(y_t):
                correct += 1
            else:
                wrong += 1
    print("正确预测%d个，正确率: %f" % (correct, correct / (correct + wrong)))
    return correct / (correct + wrong)

def main():
    # 配置参数
    epoch_num = 20  # 训练轮数
    batch_size = 20  # 每轮训练样本数
    train_num = 5000  # 每轮训练总训练样本总数
    input_size = 5  # 输入向量维度
    lr = 0.001  # 学习率

    # 建立模型
    model = torch_model(input_size)

    # 选择优化器
    optim = torch.optim.Adam(model.parameters(), lr = lr)
    log = []

    # 创建训练集，正常任务是读取训练集
    train_x, train_y = build_dataset(train_num)

    # 训练过程
    for epoch in range(epoch_num):
        model.train()
        watch_loss = []
        for batch_index in range(train_num // batch_size):
            x = train_x[batch_index * batch_size : (batch_index + 1) * batch_size]
            y = train_y[batch_index * batch_size : (batch_index + 1) * batch_size]
            loss = model(x, y)  # 计算loss model.forward(x, y)
            loss.backward()  # 计算梯度
            optim.step()  # 更新权重
            optim.zero_grad()  # 梯度归零
            watch_loss.append(loss.item())
        print("==========\n第%d轮平均loss: %f" % (epoch + 1, np.mean(watch_loss)))
        acc = evaluate(model)  # 测试本轮模型结果
        log.append([acc, float(np.mean(watch_loss))])

    # 保存模型
    torch.save(model.state_dict(), "model—week2.bin")

    # 画图
    print(log)
    plt.plot(range(len(log)), [l[0] for l in log], label="acc")
    plt.plot(range(len(log)), [l[1] for l in log], label="loss")
    plt.legend()
    plt.show()
    return

# 使用训练好的模型做预测
def predict(model_path, input_vec):
    input_size = 5
    model = torch_model(input_size)
    model.load_state_dict(torch.load(model_path))  # 加载训练权重
    print(model.state_dict())

    model.eval()  # 测试模式
    with torch.no_grad():  # 不计算梯度
        input_vec = torch.FloatTensor(np.array(input_vec))
        result = model(input_vec)  # 模型预测
    for vec, res in zip(input_vec, result):
        pred_class = torch.argmax(res).item()
        print("输入:%s, 预测类别:%d, 概率值:%f" % (vec.numpy(), pred_class, res[pred_class]))

if __name__ == "__main__":
    main()
